In [5]:
# Import the modules and the required dependencies
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

# Supress Warnings
import warnings
warnings.simplefilter(action = "ignore")

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Credit Risk` folder into a Pandas DataFrame.

In [6]:
# Read the CSV file from the Credit Risk folder into a Pandas DataFrame
data = Path("./lending_data.csv")
df_lending_data = pd.read_csv(data)

# Define a dictionary structure for header row styling
df_header_row = {
    'selector': 'th.col_heading',
    'props': 'background-color: bisque;'
}

# Review the DataFrame and apply styling appropriately
df_lending_data.head().style.hide().set_table_styles([df_header_row])

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [7]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending_data['loan_status']

# Separate the X variable, the features
X= df_lending_data.drop(columns=['loan_status'])

In [8]:
# Review the y variable Series
display(y.head())
display(y.tail())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, dtype: int64

In [9]:
# Review the X variable DataFrame and apply styling appropriately
display(X.head().style.hide().set_table_styles([df_header_row]))
display(X.tail().style.hide().set_table_styles([df_header_row]))

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [10]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [11]:
# Check the balance of our target values
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

In [12]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(solver='lbfgs',max_iter=200,random_state=1)

# Fit the model using training data
logistic_regression_model.fit(X_train, y_train)

LogisticRegression(max_iter=200, random_state=1)

### Step 2: Save predictions on testing data labels by using testing feature data (`X_test`) and the fitted model.

In [13]:
# Make a prediction using the testing data
testing_data_predictions = logistic_regression_model.predict(X_test)
results = pd.DataFrame({"Prediction": testing_data_predictions, "Actual": y_test}).reset_index(drop=True)

# Display the first few rows of the dataframe with relevant styling
results.head().style.hide().set_table_styles([df_header_row])

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [14]:
# Generate a confusion matrix for the model
testing_matrix_model = confusion_matrix(y_test, testing_data_predictions)
testing_matrix_model_df = pd.DataFrame(
    testing_matrix_model, columns=["Predicted 0", "Predicted 1"]
)

# Display the confusion matrix with relevant styling
display(testing_matrix_model_df.style.hide().set_table_styles([df_header_row]))

In [15]:
# Print the classification report for the model
print("-------------------------\n")
print("Classification Report")
print("-------------------------\n")
target_names = ["healthy loan","high-risk loan"]
print(classification_report(y_test, testing_data_predictions, target_names = target_names))

-------------------------

Classification Report
-------------------------

                precision    recall  f1-score   support

  healthy loan       1.00      1.00      1.00     18759
high-risk loan       0.87      0.89      0.88       625

      accuracy                           0.99     19384
     macro avg       0.94      0.94      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:**

1) Precision measures the proportion of correctly predicted instances out of the total predicted instances for a specific label. For the *healthy loan* class, the precision is 1.00, indicating that the model predicts this label with perfect precision. For the *high-risk loan* class, the precision is 0.87, meaning that 87% of the predicted high-risk loans are correct.

2) Recall, also known as sensitivity or true positive rate, measures the proportion of correctly predicted instances out of the total actual instances for a specific label. The recall for the *healthy loan* class is 1.00, indicating that the model captures all the actual healthy loans. The recall for the *high-risk loan* class is 0.89, meaning that the model identifies 89% of the high-risk loans correctly.

3) The F1-score is the mean of precision and recall and provides a balanced measure of the model's performance. For the *healthy loan* class, the F1-score is 1.00, indicating perfect performance. The F1-score for the *high-risk loan* class is 0.88, representing a good balance between precision and recall for this label.

Based on the above accuracy score of 99%, this model performs exceptionally well.

